### PR TITLE
RemoteDeltaLog accept input Metadata

### PR DESCRIPTION
**Pull request by linzhou-db**
RemoteDeltaLog can be initialized with `initDeltaTableMetadata`.

### CODE REVIEW

File: client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 1]
```diff
@@ -41,17 +41,19 @@ import io.delta.sharing.spark.perf.DeltaSharingLimitPushDown
 private[sharing] class RemoteDeltaLog(
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:41]: Consider adding comments on how the init metadata is used.

File: client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 2]
```diff
@@ -165,7 +169,9 @@ class RemoteSnapshot(
   }
 
   private def getTableMetadata: (Metadata, Protocol, Long) = {
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:171]: Consider adding some comments about initDeltaTableMetadata?

File: client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala [hunk 3]
```diff
@@ -165,7 +169,9 @@ class RemoteSnapshot(
...
-    val tableMetadata = client.getMetadata(table, versionAsOf, timestampAsOf)
+    val tableMetadata = initDeltaTableMetadata.getOrElse(
+      client.getMetadata(table, versionAsOf, timestampAsOf)
+    )
```
Comment (chakankardb) [inline client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:174]: Question: 
Should we cache the metadata as well?

We could have a local var that can be initialzed from initDeltaTableMetadata and update it if not set, which will avoid a refetch across multiple calls even if initDeltaTableMetadata is not specified.

Comment (linzhou-db) [reply client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala]: updated. using val as there's no re-assignment to it i the code.

File: client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 2 above)
Comment (linzhou-db) [reply client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala]: added.

File: client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala (hunk 1 above)
Comment (linzhou-db) [reply client/src/main/scala/io/delta/sharing/spark/RemoteDeltaLog.scala:49]: added

File: spark/src/test/scala/io/delta/sharing/spark/RemoteDeltaLogSuite.scala [hunk 4]
```diff
@@ -505,4 +491,34 @@ class RemoteDeltaLogSuite extends SparkFunSuite with SharedSparkSession {
...
+
+  test("RemoteDeltaLog Initialized with metadata") {
+    val path = new Path("profileFile")
+    val table = Table(share = "share", schema = "schema", name = "table")
+    val client = new TestDeltaSharingClient()
+    client.clear()
+
+    def checkGetMetadataCalledOnce(versionAsOf: Option[Long] = None, nullable: Boolean): Unit = {
+      var deltaTableMetadata = client.getMetadata(table, versionAsOf, None)
+      assert(TestDeltaSharingClient.numMetadataCalled == 1)
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/RemoteDeltaLogSuite.scala:503]: Could we test that even if create the log without deltaTableMetadata specified, getMetadata will be only called once due to caching?

**Review by chakankardb — APPROVED** [review]

File: spark/src/test/scala/io/delta/sharing/spark/RemoteDeltaLogSuite.scala (hunk 4 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/RemoteDeltaLogSuite.scala:509]: tested.

**Pull request merged**